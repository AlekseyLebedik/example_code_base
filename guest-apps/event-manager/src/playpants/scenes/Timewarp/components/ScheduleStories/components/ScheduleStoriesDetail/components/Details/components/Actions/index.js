import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import Grid from '@material-ui/core/Grid';
import { joinPath } from 'dw/core/helpers/path';
import { getBaseURL } from 'playpants/components/App/selectors';
import ConfirmActionComponent from 'dw/core/components/ConfirmActionComponent';
import IconButton from 'dw/core/components/IconButton';
import styles from './index.module.css';
import * as actions from './actions';

const DeleteScheduleStoryComponent = ({ eventCount, isSafe, onClick }) =>
  isSafe.toDelete ? (
    <ConfirmActionComponent
      className={styles.secondaryButton}
      component="IconButton"
      confirm={{
        title: 'Confirm Story Deletion',
        confirmMsg: `You are about to destroy this story and all the existing events (${eventCount}) on this story. Do you want to continue?`,
        mainButtonLabel: 'Delete',
        destructive: true,
      }}
      onClick={onClick}
      tooltip="Delete Story"
      cancelOnBackdropClick
    >
      delete_forever
    </ConfirmActionComponent>
  ) : (
    <IconButton tooltip={isSafe.unsafeReason} icon="delete_forever" disabled />
  );

DeleteScheduleStoryComponent.propTypes = {
  eventCount: PropTypes.number.isRequired,
  isSafe: PropTypes.shape({
    toDelete: PropTypes.bool,
    unsafeReason: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
};

export const ActionsBase = ({
  eventCount,
  onDeleteThenRedirect,
  storyId,
  isSafe,
}) => (
  <div className={styles.actionPanel}>
    <Grid container justify="flex-end" spacing={0} className={styles.grid}>
      <Grid item>
        <DeleteScheduleStoryComponent
          styles={styles}
          onClick={() => onDeleteThenRedirect(storyId)}
          isSafe={isSafe}
          eventCount={eventCount}
        />
      </Grid>
    </Grid>
  </div>
);

ActionsBase.propTypes = {
  isSafe: PropTypes.shape({ toDelete: PropTypes.bool }).isRequired,
  eventCount: PropTypes.number.isRequired,
  onDeleteThenRedirect: PropTypes.func.isRequired,
  storyId: PropTypes.number.isRequired,
};

const mapStateToProps = state => ({
  baseUrl: getBaseURL(state),
});

const mapDispatchToProps = dispatch => ({
  onDeleteThenRedirect: bindActionCreators(
    actions.deleteThenRedirect,
    dispatch
  ),
});

const mergeProps = (propsFromState, propsFromDispatch, ownProps) => ({
  ...propsFromState,
  ...propsFromDispatch,
  ...ownProps,
  onDeleteThenRedirect: storyId =>
    propsFromDispatch.onDeleteThenRedirect(storyId, () =>
      ownProps.history.push(
        joinPath(propsFromState.baseUrl, 'timewarp', 'schedules')
      )
    ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
  mergeProps
)(ActionsBase);
