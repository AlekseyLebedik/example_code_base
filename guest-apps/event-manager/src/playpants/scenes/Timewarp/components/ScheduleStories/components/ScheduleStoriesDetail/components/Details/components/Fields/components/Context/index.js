import React from 'react';
import PropTypes from 'prop-types';

import TextField from '@material-ui/core/TextField';

const Context = ({ context }) => (
  <TextField
    disabled
    label="Context"
    fullWidth
    value={context}
    InputProps={{
      readOnly: true,
    }}
  />
);

Context.propTypes = {
  context: PropTypes.string,
};

Context.defaultProps = {
  context: '---',
};

export default Context;
