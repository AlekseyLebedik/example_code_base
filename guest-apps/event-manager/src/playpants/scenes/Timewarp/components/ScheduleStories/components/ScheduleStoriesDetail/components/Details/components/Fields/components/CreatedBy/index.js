import React from 'react';
import PropTypes from 'prop-types';

import TextField from '@material-ui/core/TextField';

const CreatedBy = ({ createdBy }) => (
  <TextField
    disabled
    label="Created By"
    fullWidth
    value={createdBy ? createdBy.name : '---'}
    InputProps={{
      readOnly: true,
    }}
  />
);

CreatedBy.propTypes = {
  createdBy: PropTypes.object,
};

CreatedBy.defaultProps = {
  createdBy: null,
};

export default CreatedBy;
