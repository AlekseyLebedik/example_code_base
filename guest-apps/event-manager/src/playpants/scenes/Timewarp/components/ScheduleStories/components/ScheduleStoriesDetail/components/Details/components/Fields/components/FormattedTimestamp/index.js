import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  formatDateTime,
  timezoneOrDefaultSelector,
} from 'playpants/helpers/dateTime';
import TextField from '@material-ui/core/TextField';

export const FormattedTimestampBase = ({ timestamp, timezone, label }) => {
  const formattedTimestamp = timestamp
    ? formatDateTime(timestamp, undefined, timezone)
    : '---';
  return (
    <TextField
      disabled
      label={label}
      fullWidth
      value={formattedTimestamp}
      InputProps={{
        readOnly: true,
      }}
    />
  );
};

FormattedTimestampBase.propTypes = {
  timestamp: PropTypes.number,
  label: PropTypes.string.isRequired,
  timezone: PropTypes.string.isRequired,
};

FormattedTimestampBase.defaultProps = {
  timestamp: null,
};

const mapStateToProps = state => ({
  timezone: timezoneOrDefaultSelector(state),
});

export default connect(mapStateToProps)(FormattedTimestampBase);
