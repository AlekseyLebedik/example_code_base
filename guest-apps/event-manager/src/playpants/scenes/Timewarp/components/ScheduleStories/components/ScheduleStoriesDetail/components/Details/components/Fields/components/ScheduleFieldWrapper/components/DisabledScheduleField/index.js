import React from 'react';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import Tooltip from '@material-ui/core/Tooltip';

const DisabledScheduleField = ({ schedule, disabledTooltip }) => (
  <Tooltip title={disabledTooltip}>
    <TextField
      disabled
      label="Schedule"
      fullWidth
      value={schedule}
      InputProps={{
        readOnly: true,
      }}
    />
  </Tooltip>
);

DisabledScheduleField.propTypes = {
  schedule: PropTypes.string,
  disabledTooltip: PropTypes.string.isRequired,
};

DisabledScheduleField.defaultProps = {
  schedule: '---',
};

export default DisabledScheduleField;
