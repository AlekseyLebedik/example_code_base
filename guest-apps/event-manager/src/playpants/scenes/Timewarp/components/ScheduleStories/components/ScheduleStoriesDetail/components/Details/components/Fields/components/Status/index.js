import React from 'react';
import startCase from 'lodash/startCase';
import PropTypes from 'prop-types';

import TextField from '@material-ui/core/TextField';

const Status = ({ status }) => (
  <TextField
    disabled
    label="Status"
    fullWidth
    value={startCase(status)}
    InputProps={{
      readOnly: true,
    }}
  />
);

Status.propTypes = {
  status: PropTypes.string,
};

Status.defaultProps = {
  status: '---',
};

export default Status;
