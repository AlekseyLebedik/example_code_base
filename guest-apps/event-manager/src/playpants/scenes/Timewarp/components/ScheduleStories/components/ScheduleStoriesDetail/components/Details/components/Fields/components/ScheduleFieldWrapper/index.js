import React from 'react';
import PropTypes from 'prop-types';
import AdvancedScheduleField from 'playpants/components/StoryFormComponents/components/AdvancedScheduleField';
import DisabledScheduleField from './components/DisabledScheduleField';

const ScheduleFieldWrapper = props => {
  const { disabledTooltip, isDisabled, ...scheduleFieldProps } = props;
  const scheduleOption =
    scheduleFieldProps.options.find(
      option => option.value === scheduleFieldProps.value
    ) || {};
  return !isDisabled ? (
    <AdvancedScheduleField {...scheduleFieldProps} />
  ) : (
    <DisabledScheduleField
      schedule={scheduleOption.label}
      disabledTooltip={disabledTooltip}
    />
  );
};

ScheduleFieldWrapper.propTypes = {
  disabledTooltip: PropTypes.oneOfType([PropTypes.bool, PropTypes.string])
    .isRequired,
  isDisabled: PropTypes.bool.isRequired,
  options: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleOnLoadComplete: PropTypes.func.isRequired,
};

export default ScheduleFieldWrapper;
