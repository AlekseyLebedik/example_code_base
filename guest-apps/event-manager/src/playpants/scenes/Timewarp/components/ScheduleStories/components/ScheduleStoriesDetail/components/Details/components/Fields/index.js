import React, { useEffect, useState } from 'react';
import { Form, reduxForm, formValueSelector, submit } from 'redux-form';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import {
  globalSettingsSelector,
  scheduleOptionsSelector,
  schedulesDataSelector,
  titleEnvOptionsSelector,
  titleEnvsSelector,
} from 'playpants/components/App/selectors';
import CategoryField from 'playpants/components/StoryFormComponents/components/CategoryField';
import NameField from 'playpants/components/StoryFormComponents/components/NameField';
import DescriptionField from 'playpants/components/StoryFormComponents/components/DescriptionField';
import TitleEnvField from 'playpants/components/StoryFormComponents/components/TitleEnvField';
import ConfirmDialog from 'dw/core/components/Confirm';
import { fetchSchedules } from 'playpants/components/App/actions';
import CreatedBy from './components/CreatedBy';
import Status from './components/Status';
import Context from './components/Context';
import FormattedTimestamp from './components/FormattedTimestamp';
import ScheduleField from './components/ScheduleFieldWrapper';
import {
  categoryOptionsSelector,
  categoriesFetchDataSelector,
  selectedScheduleStoryContextNameSelector,
} from '../../selectors';
import { fetchCategories, fetchContexts } from '../../actions';
import { SCHEDULE_STORY_DETAIL_FORM } from './constants';
import * as actions from './actions';

export const FieldsBase = props => {
  const {
    categoryFetchOptions,
    categoryOptions,
    change,
    currentProject,
    eventCount,
    form,
    handleSubmit,
    initialValues,
    isSafe,
    onFetchCategories,
    onFetchContexts,
    onFetchSchedules,
    onPatchScheduleStory,
    onSetSelectedSchedule,
    onUploadStorySchedule,
    scheduleOptions,
    schedulesData,
    selectedScheduleStory,
    selectedScheduleStoryContextName,
    titleEnvOptions,
    titleEnvs,
    values,
  } = props;
  const [isConfirmScheduleChangeOpen, setConfirmScheduleChangeOpen] =
    useState(false);

  const [metaScheduleValues, setMetaScheduleValues] = useState({});
  useEffect(() => {
    onFetchSchedules(currentProject);
  }, []);
  useEffect(() => {
    if (selectedScheduleStory.title_env) {
      const titleEnv = titleEnvs.find(
        currentTitleEnv =>
          currentTitleEnv.id === selectedScheduleStory.title_env
      );
      onFetchContexts(titleEnv.titleId, titleEnv.shortType);
    }
  }, [selectedScheduleStory.title_env]);
  useEffect(() => {
    if (categoryFetchOptions.context) {
      const { titleId, env, context } = categoryFetchOptions;
      onFetchCategories(titleId, env, context);
    }
  }, [categoryFetchOptions.context]);

  const handleChange = (_event, newValue, previousValue, name) => {
    if (newValue !== previousValue || newValue !== initialValues[name]) {
      onPatchScheduleStory({ ...initialValues, [name]: newValue });
    }
  };

  const customHandleChange = (newValue, previousValue, name) => {
    setConfirmScheduleChangeOpen(true);
    setMetaScheduleValues({ newValue, previousValue, name });
  };

  const revertSchedulePrevious = () => {
    change(metaScheduleValues.name, metaScheduleValues.previousValue);
    setConfirmScheduleChangeOpen(false);
  };

  return (
    <div key={selectedScheduleStory.id}>
      <ConfirmDialog
        open={isConfirmScheduleChangeOpen}
        title="Confirm Story Schedule Change"
        content={`You are about to destroy and replace all the existing events (${eventCount}) on this story. Do you want to continue?`}
        destructive
        onConfirm={e =>
          handleChange(
            e,
            metaScheduleValues.newValue,
            metaScheduleValues.previousValue,
            metaScheduleValues.name
          )
        }
        onHide={revertSchedulePrevious}
      />
      <Form onSubmit={handleSubmit(onPatchScheduleStory)}>
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <NameField onBlur={handleChange} />
          </Grid>
          <Grid item xs={12}>
            <DescriptionField onBlur={handleChange} />
          </Grid>
          <Grid item xs={12}>
            <Status status={selectedScheduleStory.state} />
          </Grid>
          <Grid item xs={12}>
            <Context context={selectedScheduleStoryContextName} />
          </Grid>
          <Grid item xs={12}>
            <CategoryField
              allowNull={false}
              options={categoryOptions}
              onChange={handleChange}
              disabled={
                !isSafe.toEditSchedule || initialValues.category === 'All'
              }
            />
          </Grid>
          <Grid item xs={12}>
            <ScheduleField
              defaultValue={initialValues.schedule}
              formName={form}
              handleOnLoadComplete={onUploadStorySchedule}
              options={scheduleOptions}
              value={values.schedule}
              isClearable={false}
              customHandleScheduleChange={customHandleChange}
              isDisabled={!isSafe.toEditSchedule}
              disabledTooltip={isSafe.unsafeReason}
              schedulesData={schedulesData}
              onSetSelectedSchedule={onSetSelectedSchedule}
            />
          </Grid>
          <Grid item xs={12}>
            <TitleEnvField
              options={titleEnvOptions}
              onChange={handleChange}
              disabled
            />
          </Grid>
          <Grid item xs={12}>
            <FormattedTimestamp
              timestamp={selectedScheduleStory.start_at}
              label="Start At"
            />
          </Grid>
          <Grid item xs={12}>
            <FormattedTimestamp
              timestamp={selectedScheduleStory.end_at}
              label="End At"
            />
          </Grid>
          <Grid item xs={12}>
            <CreatedBy createdBy={selectedScheduleStory.created_by} />
          </Grid>
          <Grid item xs={12}>
            <FormattedTimestamp
              timestamp={selectedScheduleStory.created_at}
              label="Created At"
            />
          </Grid>
          <Grid item xs={12}>
            <FormattedTimestamp
              timestamp={selectedScheduleStory.updated_at}
              label="Updated At"
            />
          </Grid>
        </Grid>
      </Form>
    </div>
  );
};

FieldsBase.propTypes = {
  isSafe: PropTypes.shape({
    toEditSchedule: PropTypes.bool,
    unsafeReason: PropTypes.string,
  }).isRequired,
  categoryFetchOptions: PropTypes.object.isRequired,
  categoryOptions: PropTypes.arrayOf(PropTypes.object).isRequired,
  change: PropTypes.func.isRequired,
  currentProject: PropTypes.object.isRequired,
  eventCount: PropTypes.number.isRequired,
  form: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
  onFetchCategories: PropTypes.func.isRequired,
  onFetchContexts: PropTypes.func.isRequired,
  onFetchSchedules: PropTypes.func.isRequired,
  onPatchScheduleStory: PropTypes.func.isRequired,
  onSetSelectedSchedule: PropTypes.func.isRequired,
  onUploadStorySchedule: PropTypes.func.isRequired,
  scheduleOptions: PropTypes.arrayOf(PropTypes.object).isRequired,
  schedulesData: PropTypes.arrayOf(PropTypes.object).isRequired,
  selectedScheduleStory: PropTypes.object.isRequired,
  selectedScheduleStoryContextName: PropTypes.string.isRequired,
  titleEnvOptions: PropTypes.arrayOf(PropTypes.object).isRequired,
  titleEnvs: PropTypes.arrayOf(PropTypes.object).isRequired,
  values: PropTypes.object.isRequired,
};
const mapStateToProps = (state, { form }) => ({
  categoryOptions: categoryOptionsSelector(state),
  categoryFetchOptions: categoriesFetchDataSelector(state),
  globalSettings: globalSettingsSelector(state),
  values: formValueSelector(form)(
    state,
    'name',
    'description',
    'title_env',
    'schedule',
    'storyId'
  ),
  schedulesData: schedulesDataSelector(state),
  scheduleOptions: scheduleOptionsSelector(state),
  selectedScheduleStoryContextName:
    selectedScheduleStoryContextNameSelector(state),
  titleEnvOptions: titleEnvOptionsSelector(state),
  titleEnvs: titleEnvsSelector(state),
});

const mapDispatchToProps = (dispatch, { form }) => ({
  onSubmit: () => dispatch(submit(form)),
  onPatchScheduleStory: values =>
    dispatch(actions.patchScheduleStory(values, form)),
  onUploadStorySchedule: (
    scheduleFile,
    clearFileCb,
    closeDrawerCb,
    setAsDefaultCb,
    scheduleType
  ) =>
    dispatch(
      actions.uploadStorySchedule({
        scheduleType,
        scheduleFormData: scheduleFile,
        clearFileCb,
        closeDrawerCb,
        setAsDefaultCb,
      })
    ),
  onFetchSchedules: currentProject =>
    dispatch(fetchSchedules({ project: currentProject.id })),
  onFetchCategories: (titleEnvId, envShortType, context) =>
    dispatch(fetchCategories(titleEnvId, envShortType, context)),
  onFetchContexts: (titleEnvId, envShortType) =>
    dispatch(fetchContexts(titleEnvId, envShortType)),
  onSetSelectedSchedule: schedule =>
    dispatch(actions.setSelectedSchedule(form, schedule)),
});

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...stateProps,
  ...dispatchProps,
  ...ownProps,
  onUploadStorySchedule: (
    scheduleFile,
    clearFileCb,
    closeDrawerCb,
    setAsDefaultCb
  ) =>
    dispatchProps.onUploadStorySchedule(
      scheduleFile,
      clearFileCb,
      closeDrawerCb,
      setAsDefaultCb,
      stateProps.globalSettings.schedule_type
    ),
});

export default reduxForm({
  form: SCHEDULE_STORY_DETAIL_FORM,
  enableReinitialize: true,
})(connect(mapStateToProps, mapDispatchToProps, mergeProps)(FieldsBase));
