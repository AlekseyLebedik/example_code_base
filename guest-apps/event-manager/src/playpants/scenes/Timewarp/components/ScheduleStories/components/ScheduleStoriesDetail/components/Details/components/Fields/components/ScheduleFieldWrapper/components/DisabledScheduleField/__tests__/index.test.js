import React from 'react';
import { shallow } from 'enzyme';
import { disabledScheduleFieldProps } from 'playpants/testUtils/timewarp/components/scheduleStoriesProps';
import DisabledScheduleField from '../index';

describe('DisabledScheduleField', () => {
  const root = shallow(
    <DisabledScheduleField {...disabledScheduleFieldProps} />
  );
  it('renders default properly', () => {
    expect(root).toMatchSnapshot();
  });
});
