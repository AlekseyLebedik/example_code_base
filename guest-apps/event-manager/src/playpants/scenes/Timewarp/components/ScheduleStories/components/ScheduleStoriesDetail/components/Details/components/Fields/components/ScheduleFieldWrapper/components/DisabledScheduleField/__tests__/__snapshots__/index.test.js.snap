// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisabledScheduleField renders default properly 1`] = `
<WithStyles(ForwardRef(Tooltip))
  title=""
>
  <WithStyles(ForwardRef(TextField))
    InputProps={
      Object {
        "readOnly": true,
      }
    }
    disabled={true}
    fullWidth={true}
    label="Schedule"
    value=""
  />
</WithStyles(ForwardRef(Tooltip))>
`;
